utils/man/hosp-set.1
.TH "hosp-set" "1" "2024-03-08" "hosp" "ODROID Smart Power Utilities"
.SH "NAME"
.LP
hosp\-set \- set an ODROID Smart Power ON/OFF and START/STOP status
.SH "SYNPOSIS"
.LP
\fBhosp\-set\fP
\fIOPTION\fP
[\fIOPTION\fP]...
.SH "DESCRIPTION"
.LP
Put an ODROID Smart Power into the desired ON/OFF and START/STOP state.
.LP
ON/OFF controls whether power is provided to the system attached to the device.
Turning on from an off state starts the power flow, but the meters are not started.
.LP
START/STOP controls the current, power, and energy meters.
Starting from a stopped state starts the meters.
Stopping from a started state pauses the energy meter, but the current and power meters continue running.
Starting from a paused state resets the energy meter.
.LP
Requesting to start the device meters while in an off state has no effect.
.SH "OPTIONS"
.LP
.TP
\fB\-h\fP, \fB\-\-help\fP
Prints the help screen.
.TP
\fB\-p\fP, \fB\-\-path\fP
Device path (defaults to the first Smart Power found).
.TP
\fB\-o\fP, \fB\-\-onoff=0|1\fP
Turn the device ON (1) or OFF (0)
.TP
\fB\-s\fP, \fB\-\-startstop=0|1\fP
START (1) or STOP (0) the device meters
.SH "EXAMPLES"
.TP
\fBhosp\-set \-o 1\fP
Turn the device on if not already turned on.
.TP
\fBhosp\-set \-p /dev/hidraw1 \-o 1\fP
Turn the device /dev/hidraw1 on if not already turned on.
.TP
\fBhosp\-set \-o 0\fP
Turn the device off if not already turned off.
.TP
\fBhosp\-set \-s 1\fP
Start the device meters if not already started.
.TP
\fBhosp\-set \-s 0\fP
Stop the device meters if not already stopped.
.TP
\fBhosp\-set \-o 1 \-s 1\fP
Turn the device on if not already on and start the meters if not already started.
.TP
\fBhosp\-set \-o 1 \-s 0\fP
Turn the device on if not already on and stop/pause the meters if already started.
.SH "BUGS"
.LP
Report bugs upstream at <https://github.com/energymon/hosp>
.SH "SEE ALSO"
.LP
\fBhosp\-enumerate\fP(1), \fBhosp\-get\fP(1), \fBhosp\-poll\fP(1)

RELEASES.md
# Release Notes

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]

### Added

- hosp_enumerate: function to wrap `hid_enumerate`.
- hosp_open_device: function to open using an existing `hid_device`.
- hosp_get_device: function to get the underlying `hid_device` handle.
- hosp-enumerate: new executable to find and print ODROID Smart Power device paths.
- hosp-{get,poll,set}: `-p`/`--path` CLI argument.

### Changed

- HIDAPI is now a public dependency (was private).
- Build:
  - Increase minimum required CMake version to 3.16.
  - Prefer native HIDAPI backends over `libusb` backend.
  - Add `HOSP_HIDAPI_PC_MODULES` CMake option to let user configure the HIDAPI module search order.
- README: update documentation.
- CI: Switch from Travis CI to GitHub workflows for continuous integration.

### Fixed

- hosp_request_version_read: stringop-truncation warning with gcc >= 8.
- hosp-{poll,set}: missing-noreturn warning for `print_usage` functions.
- pkg-config: pc file is broken when CMAKE_INSTALL_{INCLUDE,LIB}DIR is absolute.


## v0.1.0 - 2018-05-31

### Added

- Initial public release.

[Unreleased]: https://github.com/energymon/hosp/compare/v0.1.0...HEAD


utils/hosp-set.c
/**
 * Manage an ODROID Smart Power.
 *
 * @author Connor Imes
 * @date 2018-05-22
 */
#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <hidapi.h>
#include <hosp.h>
#include "util.h"

static const char* path = NULL;
// -1 if unset, 0 for OFF/STOP, 1 for ON/START
static int action_onoff = -1;
static int action_startstop = -1;

static const char short_options[] = "hp:o:s:";
static const struct option long_options[] = {
  {"help",      no_argument,       NULL, 'h'},
  {"path",      required_argument, NULL, 'p'},
  {"onoff",     required_argument, NULL, 'o'},
  {"startstop", required_argument, NULL, 's'},
  {0, 0, 0, 0}
};

__attribute__ ((noreturn))
static void print_usage(int exit_code) {
  fprintf(exit_code ? stderr : stdout,
          "Usage: hosp-set OPTION [OPTION]...\n"
          "Options:\n"
          "  -h, --help               Print this message and exit\n"
          "  -p, --path               Device path (defaults to the first Smart Power found)\n"
          "  -o, --onoff=1|0          Turn the device ON (1) or OFF (0)\n"
          "  -s, --startstop=1|0      START (1) or STOP (0) the device\n");
  exit(exit_code);
}

static void parse_args(int argc, char** argv) {
  int c;
  while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
    switch (c) {
      case 'h':
        print_usage(0);
        break;
      case 'p':
        path = optarg;
        break;
      case 'o':
        action_onoff = atoi(optarg) ? 1 : 0;
        break;
      case 's':
        action_startstop = atoi(optarg) ? 1 : 0;
        break;
      case '?':
      default:
        print_usage(EINVAL);
        break;
    }
  }
  if (action_onoff < 0 && action_startstop < 0) {
    fprintf(stderr, "Must specify at least one of the -o or -s options.\n");
    print_usage(EINVAL);
  }
}

static int hosp_set_status(hosp_device* hosp) {
  int is_on;
  int is_started;
  int ret = 0;
  // first get the current status
  if (hosp_util_get_status(hosp, &is_on, &is_started)) {
    ret = errno;
    perror("Failed to get status from ODROID Smart Power");
    return ret;
  }
  if (action_onoff >= 0 && action_onoff != is_on) {
    // toggle ON/OFF
    hosp_util_msleep(HOSP_WRITE_READ_DELAY_MS);
    if (hosp_request_onoff_write(hosp)) {
      ret = errno;
      perror("Failed to toggle ON/OFF");
      return ret;
    }
  }
  if (action_startstop >= 0 && action_startstop != is_started) {
    // toggle START/STOP
    hosp_util_msleep(HOSP_WRITE_READ_DELAY_MS);
    if (hosp_request_startstop_write(hosp)) {
      ret = errno;
      perror("Failed to toggle START/STOP");
      return ret;
    }
  }
  return ret;
}

int main(int argc, char** argv) {
  hid_device* hdev = NULL;
  hosp_device* hosp;
  int ret;

  parse_args(argc, argv);

  if (hid_init() < 0) {
    fprintf(stderr, "hid_init: %ls\n", hid_error(NULL));
    return 1;
  }

  if (path != NULL) {
    if ((hdev = hid_open_path(path)) == NULL) {
      fprintf(stderr, "%ls\n", hid_error(NULL));
      ret = 1;
      goto exit_hid;
    }
  }

  if ((hosp = hosp_open_device(hdev)) == NULL) {
    perror("Failed to open ODROID Smart Power connection");
    ret = errno;
    goto close_hdev;
  }

  ret = hosp_set_status(hosp);

  if (hosp_close(hosp)) {
    ret = errno;
    perror("Failed to close ODROID Smart Power connection");
  }

close_hdev:
  if (hdev != NULL) {
    hid_close(hdev);
  }

exit_hid:
  hid_exit();
  return ret;
}
